File: core/src/commonMain/kotlin/work/socialhub/kmastodon/api/ListsResource.java
package work.socialhub.kmastodon.api;

import mastodon4j.entity.Account;
import mastodon4j.entity.List;
import mastodon4j.entity.share.Response;

/**
 * @author uakihir0
 */
public interface ListsResource {

    /**
     * Fetching a user's lists.
     *
     * @return an array of Lists created by the authenticated user
     */
    public Response<List[]> getLists();

    /**
     * Fetching the user's lists that a given account is part of.
     *
     * @param id the account's id
     * @return an array of Lists created by the target user
     */
    public Response<List[]> getLists(String id);

    /**
     * Fetching accounts that are in a given list.
     *
     * @param id    the target list's id
     * @param limit maximum number of results. (default: 40)
     * @return an array of Accounts that are in a specified list.
     */
    public Response<Account[]> getListAccounts(String id, Long limit);

    /**
     * Fetching list.
     *
     * @param id the target list's id
     * @return a specified list.
     */
    public Response<List> getList(String id);

    /**
     * Create list.
     *
     * @param title list name
     * @return created list
     */
    public Response<List> createList(String title);

    /**
     * Update list.
     *
     * @param id    the target list's id
     * @param title list name
     * @return updated list
     */
    public Response<List> updateList(String id, String title);

    /**
     * Delete list.
     *
     * @param id the target list's id
     */
    public void deleteList(String id);

    /**
     * Add account to list.
     *
     * @param id         the target list's id
     * @param accountIds account ids to add
     */
    public void addAccountsToList(String id, long[] accountIds);

    /**
     * Delete account to list.
     *
     * @param id         the target list's id
     * @param accountIds account ids to add
     */
    public void deleteAccountsToList(String id, long[] accountIds);
}

File: core/src/commonMain/kotlin/work/socialhub/kmastodon/api/ListsResource.kt
package work.socialhub.kmastodon.api

import work.socialhub.kmastodon.api.request.lists.ListsAddAccountsToListRequest
import work.socialhub.kmastodon.api.request.lists.ListsCreateListRequest
import work.socialhub.kmastodon.api.request.lists.ListsDeleteAccountsToListRequest
import work.socialhub.kmastodon.api.request.lists.ListsDeleteListRequest
import work.socialhub.kmastodon.api.request.lists.ListsListAccountsRequest
import work.socialhub.kmastodon.api.request.lists.ListsListRequest
import work.socialhub.kmastodon.api.request.lists.ListsListsRequest
import work.socialhub.kmastodon.api.request.lists.ListsUpdateListRequest
import work.socialhub.kmastodon.api.response.Response
import work.socialhub.kmastodon.api.response.ResponseUnit
import work.socialhub.kmastodon.api.response.lists.ListsCreateListResponse
import work.socialhub.kmastodon.api.response.lists.ListsListAccountsResponse
import work.socialhub.kmastodon.api.response.lists.ListsListResponse
import work.socialhub.kmastodon.api.response.lists.ListsListsResponse
import work.socialhub.kmastodon.api.response.lists.ListsUpdateListResponse
import kotlin.js.JsExport

@JsExport
interface ListsResource {

    /**
     * Fetching the user's lists that a given account is part of.
     */
    suspend fun lists(
        request: ListsListsRequest
    ): Response<Array<ListsListsResponse>>

    @JsExport.Ignore
    fun listsBlocking(
        request: ListsListsRequest
    ): Response<Array<ListsListsResponse>>

    /**
     * Fetching accounts that are in a given list.
     */
    suspend fun listAccounts(
        request: ListsListAccountsRequest
    ): Response<Array<ListsListAccountsResponse>>

    @JsExport.Ignore
    fun listAccountsBlocking(
        request: ListsListAccountsRequest
    ): Response<Array<ListsListAccountsResponse>>

    /**
     * Fetching list.
     */
    suspend fun list(
        request: ListsListRequest
    ): Response<ListsListResponse>

    @JsExport.Ignore
    fun listBlocking(
        request: ListsListRequest
    ): Response<ListsListResponse>

    /**
     * Create list.
     */
    suspend fun createList(
        request: ListsCreateListRequest
    ): Response<ListsCreateListResponse>

    @JsExport.Ignore
    fun createListBlocking(
        request: ListsCreateListRequest
    ): Response<ListsCreateListResponse>

    /**
     * Update list.
     */
    suspend fun updateList(
        request: ListsUpdateListRequest
    ): Response<ListsUpdateListResponse>

    @JsExport.Ignore
    fun updateListBlocking(
        request: ListsUpdateListRequest
    ): Response<ListsUpdateListResponse>

    /**
     * Delete list.
     */
    suspend fun deleteList(
        request: ListsDeleteListRequest
    ): ResponseUnit

    @JsExport.Ignore
    fun deleteListBlocking(
        request: ListsDeleteListRequest
    ): ResponseUnit

    /**
     * Add account to list.
     */
    suspend fun addAccountsToList(
        request: ListsAddAccountsToListRequest
    ): ResponseUnit

    @JsExport.Ignore
    fun addAccountsToListBlocking(
        request: ListsAddAccountsToListRequest
    ): ResponseUnit

    /**
     * Delete account to list.
     */
    suspend fun deleteAccountsToList(
        request: ListsDeleteAccountsToListRequest
    ): ResponseUnit

    @JsExport.Ignore
    fun deleteAccountsToListBlocking(
        request: ListsDeleteAccountsToListRequest
    ): ResponseUnit
}
